Adds RapidPro as an SMS Gateway

dianabarsan opened the pull request:
# Description

Adds RapidPro controller to receive messages from RapidPro instances.
Adds RapidPro service for relaying messages to RapidPro (via the `broadcast` endpoint) and poll for status updates (via the `messages` endpoint). 

medic/cht-core#6532

# Code review checklist
<!-- Remove or comment out any items that do not apply to this PR; in the remaining boxes, replace the [ ] with [x]. -->
- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

dianabarsan commented:
hey @garethbowen . Could you please review? Thanks for your time!

garethbowen commented on api/src/services/rapid-pro.js:
```diff
@@ -0,0 +1,240 @@
+      .catch(err => {
+        logger.error('Error while polling message states: %o', err);
+      });
+  },
```
If I'm reading this correctly then the most statuses we can update is 25 per minute. If we have a project with more throughput than that it'll never catch up. I think we need to loop through all non-final messages with every poll call so we can go as fast as possible.

This will exacerbate another issue where `poll` may get called before the previous `poll` has completed which could cause conflicts or at least wasted effort. The naive solution would be to just have a `running` flag. A more nuanced approach would be to not schedule the next poll call until the first one returns.

dianabarsan replied:
> I think we need to loop through all non-final messages with every poll call so we can go as fast as possible.

I actually implemented this in one iteration. Exactly with a `running` flag. I can rollback to that :) 
I'll have to remember why I went with the 25/minute, there was a reason but it eludes me now.

dianabarsan replied:
I think the reason why I didn't want to keep recursively calling the messages endpoint was because i was worried of the rapidpro request/hour limit and what happens if the limit is hit.

dianabarsan replied:
My biggest concern, I think, was if someone switched from another outgoing message service to rapid-pro, and they had a bunch of older messages that are in non-final states. 
I was wondering if we could add another property to the message with the outgoing message service that was used to generate the "gateway_ref" and have the new view index by that value and the state. 
But then I thought it was overkill. 

Now I'm unsure again. Please share your thoughts about this.

garethbowen commented on api/src/services/rapid-pro.js:
```diff
@@ -0,0 +1,240 @@
+  const configuredUrl = smsSettings &&
+                        smsSettings.rapidPro &&
+                        smsSettings.rapidPro.url;
+  const textitUrl = 'https://textit.in';
```
I'm not sure about adding a default value in here. Frankly for PII reasons we'd rather everyone used our RapidPro server than TextIt. I would require it to be configured manually.

dianabarsan replied:
Ok cool, I'll remove!

garethbowen commented on ddocs/medic-db/medic-sms/views/gateway_messages_by_state/map.js:
```diff
@@ -0,0 +1,22 @@
+  };
+  _emit(doc.tasks || []);
+  _emit(doc.scheduled_tasks || []);
+}
```
It's a shame the `medic/messages_by_state` view can't be retrofitted to do this too...

dianabarsan replied:
Yea. `messages_by_state` is used for exporting (with skip and limit and no other filters), so adding an additional emit per message would effectively duplicate every message in the export :( Oh, I think this already happens because of the 'pending-or-forwarded' additional emit, but only for messages in that state. Oh well. 

We also use `messages_by_state` with start_key + end_key, so emitting a similar additional key would produce duplicates there too, so I'd have to keep a simple, a non-array key, like this new view. 

It would have just made  `messages_by_state` too bloated IMO.

garethbowen reviewed (changes requested):
Awesome solution.

I've got some suggestions inline, but grab me on Slack if you want to talk through it!

dianabarsan commented on api/src/controllers/rapid-pro.js:
```diff
@@ -0,0 +1,56 @@
+    return Promise.reject({ code: 400, message: 'Request body is required' });
+  }
+
+  const outgoingMessageService = messaging.getOutgoingMessageService();
+  if (!outgoingMessageService || outgoingMessageService.name !== rapidPro.name) {
+    return Promise.reject({ code: 400, message: 'Service not enabled' });
+  }
```
My idea was to stop requests to this endpoint (the api route is enabled) that a bot can do, and check the validity of the token every time. I wanted to stop these requests early, when the service wasn't even enabled.

dianabarsan commented on api/src/services/messaging.js:
```diff
@@ -183,8 +197,12 @@ const resolveMissingUuids = changes => {
 
-module.exports = {
+const countStateChanges = (results, stateChangesByDocId) => {
+  return results.reduce((count, result) => count + stateChangesByDocId[result.id].length, 0);
```
I think reduce still has a place in this world, exactly when doing these operations.
I think the _sum solution is even harder to read than the reduce and does an additional iteration.

dianabarsan commented on api/src/controllers/rapid-pro.js:
```diff
@@ -0,0 +1,56 @@
+  return getIncomingToken().then(expected => {
+    // todo Reviewer: should we mimic the RapidPro authorization header, and include "Token" prefix?
+    // or use a query key like africas-talking?
+    // Consistent with our other APIs or consistent with the api integrating with?
```
My question was mostly: the outgoing part of the integration (which we control 100% through code, and requires using auth tokens) is not visible to the app developer. My question was: does someone integrating with CHT-Core with RapidPro need to know how RapidPro handles authentication.

dianabarsan commented on api/src/services/rapid-pro.js:
```diff
@@ -0,0 +1,240 @@
+    // get the credentials every call so changes can be made without restarting api
+    return getCredentials()
+      .then(credentials => {
+        const viewOptions = { keys: _.uniq(nonFinalStates), limit: BATCH_SIZE, skip };
```
Yea, absolutely, I moved code around at some point and missed this!

dianabarsan commented on api/src/controllers/rapid-pro.js:
```diff
@@ -0,0 +1,56 @@
+module.exports = {
+  // todo Reviewer: should we even have this endpoint?
+  // I added it so it would simplify the webhook call from rapidpro -> cht-core
+  // but the sms-gateway endpoint can still be used just fine.
```
Thanks!

garethbowen reviewed (approved):
Nice one! I think this is ready for AT but I'll leave it for you to decide based on some ongoing conversations.

dianabarsan commented:
hey @garethbowen . 
I pushed two more commits
- one to consolidate the "RapidPro" name (and never using a dash between rapid and pro). 
- one to handle rate limiting when polling (backing off when rate limiting is triggered). 

About rate limiting: 
- it's the standard Django rate limiting
- number of requests is per endpoint
- requesting after limit is reached doesn't do anything, except fail the request
- the `retry-after` is useless. it returns a rough idea of when the next opening spot for one request will come. For our endpoint, if the limit is reached, the maximum retry-after is 60 seconds, which is exactly the interval at which we request again. 

I've also added the documentation PR. 
I've also rebased to get the nicer error logging: 
![image](https://user-images.githubusercontent.com/35681649/113315741-46200800-9316-11eb-847f-974b2dda1405.png)

garethbowen commented:
@dianabarsan Looks great! Thanks for the polish. It'd be worth adding one unit test for the throttling code, but otherwise good to go.

dianabarsan commented:
@garethbowen 
I have an e2e test: https://github.com/medic/cht-core/pull/6995/files#diff-54f699aa61c171bd4d9c34414ed963c060ecdcd9655aeabb2de2032c8d1a93d2R575 
and a unit test: https://github.com/medic/cht-core/pull/6995/files#diff-bba8c37bbc4b7704a78002d227bfc0c5c3d2bbfa1c405d1b1b0b1de27351905dR572

Were you thinking of something specific?

garethbowen commented:
@dianabarsan Perfect! I did a ctrl+f for `429` but I suspect it missed because GH had collapsed the test files because they were too large. Sorry about that! You're good to go.

dianabarsan commented:
Thanks @garethbowen ! 

> GH had collapsed the test files 

I hate that. I wish there was an expand all button at the top.

craig-landry commented:
> wish there was an expand all button at the top

Their css is pretty simple on that stuff so a bookmarklet can work pretty well:
```
javascript:(function() { document.querySelectorAll('.load-diff-button').forEach( n => n.click()); })();
```

newtewt commented on api/src/controllers/rapidpro.js:
```diff
@@ -0,0 +1,65 @@
+  return getConfiguredIncomingToken().then(expected => {
+    if (expected !== givenToken) {
+      logger.warn(`Attempt to access RapidPro endpoint without the correct incoming token`);
+      return Promise.reject({ code: 403, message: `Incorrect token: "${givenToken}"` });
```
Should we be giving back a token even if it is incorrect?

dianabarsan replied:
Good point, maybe not?
